### PR TITLE
Proof of concept for Single Sign On

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
by @David-Development 
Ref: https://github.com/David-Development/ownCloud-Account-Importer/issues/3

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky @David-Development any change to get this updated to latest master? This would be a nice post-2.0.0 addition :)

cc @mario

**Comment by AndyScherzinger** [comment]
Is anyone able to bring this on track with actual master? AccountManagerService is quite the merge conflict :/

**Comment by AndyScherzinger** [comment]
❗️ rebased to latest master

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you kill your codacy project link please? --> https://www.codacy.com/app/tobias_2/android/pullRequest?prid=729947 (since this gets referenced in this PR instead of the Nextcloud orga link...)

**Comment by tobiasKaminsky** [comment]
Removed

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+/**
+ * Created by david on 29.06.17.
+ *
+ * More information here: https://github.com/abeluck/android-streams-ipc
+ */
+
+public class InputStreamBinder extends IInputStreamService.Stub {
+    private final static String TAG = "InputStreamBinder";
+
+    private ArrayList<String> validPackages = new ArrayList<>(Arrays.asList("de.luhmer.owncloudnewsreader"));
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: should be just `List`

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            exception = e;
+        }
+
+        try {
+            // Write exception to the stream followed by the actual network stream
+            InputStream exceptionStream = serializeObjectToInputStream(exception);
+            InputStream resultStream = new java.io.SequenceInputStream(exceptionStream, httpStream);
+            return ParcelFileDescriptorUtil.pipeFrom(resultStream, thread -> Log.d(TAG, "Done sending result"));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: should be a log statement

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+            public int read() {
+                return 0;
+            }
+        };
+        try {
+            // Start request and catch exceptions
+            NextcloudRequest request = deserializeObjectAndCloseStream(is);
+            httpStream = processRequest(request);
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: should be a log statement

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            case "GET":
+                method = new GetMethod(request.url);
+                break;
+
+            case "POST":
+                method = new PostMethod(request.url);
+                if (request.requestBody != null) {
+                    StringRequestEntity requestEntity = new StringRequestEntity(
+                            request.requestBody,
+                            "application/json",
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: "application/json" and "UTF-8" should be constants

File: src/main/java/com/nextcloud/android/sso/aidl/IThreadListener.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.nextcloud.android.sso.aidl;
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/aidl/IThreadListener.java]: missing license header

File: src/main/java/com/nextcloud/android/sso/aidl/NextcloudRequest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
+package com.nextcloud.android.sso.aidl;
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/aidl/NextcloudRequest.java]: missing license header

File: src/main/java/com/nextcloud/android/sso/aidl/NextcloudRequest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            ncr.method = method;
+            return this;
+        }
+
+        public Builder setHeader(Map<String, List<String>> header) {
+            ncr.header = header;
+            return this;
+        }
+
+        public Builder setParameter(HashMap<String, String> parameter) {
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/aidl/NextcloudRequest.java]: let's use `Map` instead :)

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
+package com.nextcloud.android.sso.aidl;
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: missing license header

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+import android.os.ParcelFileDescriptor;
+import android.util.Log;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class ParcelFileDescriptorUtil {
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: please declare final

[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: + private, documented constructor

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            int len;
+
+            try {
+                while ((len = mIn.read(buf)) > 0) {
+                    mOut.write(buf, 0, len);
+                }
+                mOut.flush(); // just to be safe
+            } catch (IOException e) {
+                Log.e("TransferThread", "writing failed");
+                e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: `Log`

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+            } finally {
+                try {
+                    mIn.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: `Log`

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                }
+                try {
+                    mOut.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: `Log`

File: src/main/java/com/owncloud/android/services/AccountManagerService.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
+package com.owncloud.android.services;
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/services/AccountManagerService.java]: missing license header

File: src/main/java/com/owncloud/android/services/AccountManagerService.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        }
+        return mBinder;
+    }
+
+    @Override
+    public boolean onUnbind(Intent intent) {
+        return super.onUnbind(intent);
+    }
+
+}
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/services/AccountManagerService.java]: new line

File: src/main/java/com/owncloud/android/ui/asynctasks/AsyncTaskHelper.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
+package com.owncloud.android.ui.asynctasks;
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/asynctasks/AsyncTaskHelper.java]: missing license header

File: src/main/java/com/owncloud/android/ui/asynctasks/AsyncTaskHelper.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.concurrent.Callable;
+
+/**
+ * Created by david on 27.06.17.
+ */
+
+public class AsyncTaskHelper {
+
+    @RequiresApi(api = Build.VERSION_CODES.HONEYCOMB)
+    public static <T> T executeBlockingRequest(Callable<T> callable) throws Exception {
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/ui/asynctasks/AsyncTaskHelper.java]: We already only support Android 4.x so annotation is obsolete

File: src/main/java/com/owncloud/android/ui/asynctasks/AsyncTaskHelper.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        GenericAsyncTaskWithCallable<T> at = new GenericAsyncTaskWithCallable<>(callable);
+
+        T result = at.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR).get();
```
[the remark this answers is not in the document]
Comment (David-Development) [reply src/main/java/com/owncloud/android/ui/asynctasks/AsyncTaskHelper.java]: Does anyone know what's wrong with this line? I'm not sure how to fix..

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+            setDaemon(true);
+        }
+
+        @Override
+        public void run() {
+            byte[] buf = new byte[1024];
+            int len;
+
+            try {
+                while ((len = mIn.read(buf)) > 0) {
```
[the remark this answers is not in the document]
Comment (David-Development) [reply src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: I tried unfolding this code snippet somehow.. but I can't find a better way to write it than as it is (without writing 10more lines). @tobiasKaminsky Any ideas? :)

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+    public InputStreamBinder(Context context) {
+        this.context = context;
+    }
+
+    private NameValuePair[] convertMapToNVP(Map<String, String> map) {
+        NameValuePair[] nvp = new NameValuePair[map.size()];
+        int i = 0;
+        for (String key : map.keySet()) {
+            nvp[i] = new NameValuePair(key, map.get(key));
```
[the remark this answers is not in the document]
Comment (David-Development) [reply src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: I think we could use an method to work around this issue.. however this doesn't make the code any better/prettier, does it?

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+
+    private InputStream processRequest(final NextcloudRequest request) throws UnsupportedOperationException, com.owncloud.android.lib.common.accounts.AccountUtils.AccountNotFoundException, OperationCanceledException, AuthenticatorException, IOException {
+        Account account = AccountUtils.getOwnCloudAccountByName(context, request.accountName); // TODO handle case that account is not found!
+        if(account == null) {
+            throw new IllegalStateException("CE_2"); // Custom Exception 2 (Account not found)
+        }
+
+        OwnCloudClientManager ownCloudClientManager = OwnCloudClientManagerFactory.getDefaultSingleton();
+        OwnCloudAccount ocAccount = new OwnCloudAccount(account, context);
+        OwnCloudClient client = ownCloudClientManager.getClientFor(ocAccount, context);
```
[the remark this answers is not in the document]
Comment (David-Development) [reply src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: Not sure what's wrong with this statement.. Any ideas? Would it help if we change the ordering to the following? 

```java
OwnCloudAccount ocAccount = new OwnCloudAccount(account, context);
OwnCloudClientManager ownCloudClientManager = OwnCloudClientManagerFactory.getDefaultSingleton();
OwnCloudClient client = ownCloudClientManager.getClientFor(ocAccount, context);
```

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+
+    public InputStreamBinder(Context context) {
+        this.context = context;
+    }
+
+    private NameValuePair[] convertMapToNVP(Map<String, String> map) {
+        NameValuePair[] nvp = new NameValuePair[map.size()];
+        int i = 0;
+        for (String key : map.keySet()) {
+            nvp[i] = new NameValuePair(key, map.get(key));
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: I've set this one to ignore, so it'll be gone with the next run.

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (David-Development) [reply src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: @AndyScherzinger As discussed with Tobias, I think we should ignore this issue here as well.

Or does anyone has a good idea on how to rewrite this without adding another 10 lines? :)

File: src/main/java/com/nextcloud/android/sso/InputStreamBinder.java [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+        is.close();
+        ois.close();
+        return result;
+    }
+
+
+    private InputStream processRequest(final NextcloudRequest request) throws UnsupportedOperationException, com.owncloud.android.lib.common.accounts.AccountUtils.AccountNotFoundException, OperationCanceledException, AuthenticatorException, IOException {
+        Account account = AccountUtils.getOwnCloudAccountByName(context, request.accountName); // TODO handle case that account is not found!
+        if(account == null) {
+            throw new IllegalStateException("CE_2"); // Custom Exception 2 (Account not found)
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/sso/InputStreamBinder.java]: If `CE_1` and `CE_2` are fixed identifiers then they should probably declared as constants and the constants being used to make sure no typos happen etc.

File: src/main/java/com/owncloud/android/authentication/AccountAuthenticator.java [hunk 23]
```diff
@@ -155,8 +158,54 @@ public Bundle editProperties(AccountAuthenticatorResponse response, String accou
...
+                Log_OC.e(TAG, "Account not found");
+                return new Bundle();
+            }
+
+            result.putString(AccountManager.KEY_ACCOUNT_NAME,  account.name);
+            result.putString(AccountManager.KEY_ACCOUNT_TYPE,  MainApp.getAccountType(mContext));
+            result.putString(AccountManager.KEY_AUTHTOKEN,     NEXTCLOUD_SSO);
+            result.putString("username", userId);
+            result.putString("token", token);
+            result.putString("server_url", serverUrl);
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/authentication/AccountAuthenticator.java]: username, token and server_url are keys and should thus be defined as constants :)

File: src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/nextcloud/android/sso/aidl/ParcelFileDescriptorUtil.java]: set to ignore, thus be gone with the next build :)

**Comment by AndyScherzinger** [comment]
The additional 6 lint warnings should be gone as soon as the transifex bot has been run again and the branch gets rebased afterwards.

File: src/main/res/values/strings.xml [hunk 25]
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
+    <string name="permission_label">nextcloud sso</string>
+    <string name="permission_desc">nextcloud sso</string>
+    <string name="permission_group">nextcloud login</string>
```
Comment (AndyScherzinger) [inline src/main/res/values/strings.xml]: are these to bet tanslated? if not please add `translatable="false"` else transifex will pick them up for translation :)

Comment (David-Development) [reply src/main/res/values/strings.xml]: I think those strings are nowhere visible for the user. Even though I would've expected them to appear in the "Permission" settings of the app. Thanks for the hint!

**Comment by tobiasKaminsky** [comment]
@David-Development lint is complaining:

> Exported services (services which either set exported=true or contain an intent-filter and do not specify exported=false) should define a permission that an entity must have in order to launch the service or bind to it. Without this, any application can use this service.
Note: This issue has an associated quickfix operation in Android Studio and IntelliJ IDEA.

https://github.com/nextcloud/android/blob/322aee0454b4bfd84b258b006d21f9affbf00cd5/src/main/AndroidManifest.xml#L301-L305

Is this what we discussed that a permission is not possible to set? Then I can ignore this warning?

**Comment by David-Development** [comment]
Hmm.. yes. This is what we discussed in https://github.com/nextcloud/Android-SingleSignOn/issues/27. Even if we keep the permission.. any app that defines that permission will be also granted it. Or am I wrong here?

**Comment by tobiasKaminsky** [comment]
No, you are right. I just wanted to be sure.
Then I'll adjust lint-results and then merge :+1:

**Comment by tobiasKaminsky** [comment]
(rebase will be a hell, so let us merge this instead, this time…)

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger are you fine with merging? :-D

**Comment by AndyScherzinger** [comment]
:+1: perfectly fine, yes and also about not rebasing ;)

**Pull request merged**